driver/src/orderbook/streamed/block_timestamp.rs

```diff
@@ -0,0 +1,23 @@
+use crate::contracts::Web3;
+use anyhow::{Context as _, Result};
+use ethcontract::{web3::types::BlockId, H256};
+use futures::{compat::Compat01As03, future::BoxFuture, FutureExt as _};
+
+/// Helper trait to make this functionality mockable for tests.
+pub trait BlockTimestamp {
+    fn block_timestamp(&self, block_hash: H256) -> BoxFuture<Result<u64>>;
+}
+
+/// During normal operation this is implemented by Web3.
+impl BlockTimestamp for Web3 {
+    fn block_timestamp(&self, block_hash: H256) -> BoxFuture<Result<u64>> {
+        let r = Compat01As03::new(self.eth().block(BlockId::Hash(block_hash)));
+        let r = r.map(move |block| -> Result<u64> {
+            let block = block
+                .with_context(|| format!("failed to get block {}", block_hash))?
+                .with_context(|| format!("block {} does not exist", block_hash))?;
+            Ok(block.timestamp.low_u64())
+        });
+        r.boxed()
```

Comment:
nit: If the future is going to be boxed, I think the code would be easier to follow using `async { ... }.boxed()`. Then you don't need to use `map` and stuff.